File: backend/app/schemas/analytics.py
from __future__ import annotations

from pydantic import BaseModel, Field


class OverviewStats(BaseModel):
    total_transactions: int = 0
    total_alerts: int = 0
    total_cases: int = 0
    total_entities: int = 0
    fraud_rate: float = 0.0
    avg_risk_score: float = 0.0
    total_amount_processed: float = 0.0
    active_rules: int = 0


class FraudTrendPoint(BaseModel):
    date: str
    count: int
    amount: float
    avg_score: float


class FraudTrendsResponse(BaseModel):
    period: str
    data_points: list[FraudTrendPoint]


class TransactionVolumePoint(BaseModel):
    date: str
    count: int
    amount: float
    by_channel: dict[str, int] = Field(default_factory=dict)


class TransactionVolumeResponse(BaseModel):
    period: str
    data_points: list[TransactionVolumePoint]


class RiskDistributionResponse(BaseModel):
    low: int = 0
    medium: int = 0
    high: int = 0
    critical: int = 0
    total: int = 0


class TopPattern(BaseModel):
    pattern_name: str
    count: int
    percentage: float
    trend: str = "stable"


class TopPatternsResponse(BaseModel):
    patterns: list[TopPattern]


class GeoDataPoint(BaseModel):
    country_code: str
    count: int
    fraud_count: int
    total_amount: float
    fraud_rate: float


class GeoResponse(BaseModel):
    data: list[GeoDataPoint]


class ModelPerformancePoint(BaseModel):
    model_name: str
    model_type: str
    accuracy: float = 0.0
    precision: float = 0.0
    recall: float = 0.0
    f1: float = 0.0
    auc_roc: float = 0.0


class ModelPerformanceResponse(BaseModel):
    models: list[ModelPerformancePoint]


class ReportRequest(BaseModel):
    report_type: str
    date_from: str | None = None
    date_to: str | None = None
    filters: dict | None = None


class ReportResponse(BaseModel):
    report_id: str
    status: str
    download_url: str | None = None
    created_at: str

File: frontend/src/types/analytics.ts
export interface OverviewStats {
  total_transactions: number;
  total_alerts: number;
  total_cases: number;
  total_entities: number;
  fraud_rate: number;
  avg_risk_score: number;
  total_amount_processed: number;
  active_rules: number;
}

export interface FraudTrendPoint {
  date: string;
  count: number;
  amount: number;
  avg_score: number;
}

export interface FraudTrendsResponse {
  period: string;
  data_points: FraudTrendPoint[];
}

export interface TransactionVolumePoint {
  date: string;
  count: number;
  amount: number;
  by_channel: Record<string, number>;
}

export interface TransactionVolumeResponse {
  period: string;
  data_points: TransactionVolumePoint[];
}

export interface RiskDistributionResponse {
  low: number;
  medium: number;
  high: number;
  critical: number;
  total: number;
}

export interface TopPattern {
  pattern_name: string;
  count: number;
  percentage: number;
  trend: string;
}

export interface TopPatternsResponse {
  patterns: TopPattern[];
}

export interface GeoDataPoint {
  country_code: string;
  count: number;
  fraud_count: number;
  total_amount: number;
  fraud_rate: number;
}

export interface GeoResponse {
  data: GeoDataPoint[];
}

export interface ModelPerformancePoint {
  model_name: string;
  model_type: string;
  accuracy: number;
  precision: number;
  recall: number;
  f1: number;
  auc_roc: number;
}

export interface ModelPerformanceResponse {
  models: ModelPerformancePoint[];
}

export interface ReportRequest {
  report_type: string;
  date_from?: string;
  date_to?: string;
  filters?: Record<string, unknown>;
}

export interface ReportResponse {
  report_id: string;
  status: string;
  download_url?: string;
  created_at: string;
}
